xql/main.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python3
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import xarray as xr
+import typing as t
+
+from sqlglot import parse_one, exp
+
+operate = {
+    "and" : lambda a, b: a & b,
+    "or" : lambda a, b: a | b,
+    "eq" : lambda a, b: a == b,
+    "gt" : lambda a, b: a > b,
+    "lt" : lambda a, b: a < b,
+    "gte" : lambda a, b: a >= b,
+    "lte" : lambda a, b: a <= b,
+}
+
+aggregate_function_map = {
+    'avg': lambda x, y: x.mean(dim=y) if y else x.mean(),
+    'min': lambda x, y: x.min(dim=y) if y else x.mean(),
+    'max': lambda x, y: x.max(dim=y) if y else x.mean(),
+}
+
+def parse(a, b):
+    arr_type = a.dtype.name
+    if arr_type == 'float64':
+        b = np.float64(b)
+    elif arr_type == 'float32':
+        b = np.float32(b)
+    elif arr_type == 'datetime64[ns]':
+        b = np.datetime64(b)
+    return a, b
+
+def evaluate(a: xr.DataArray, b, operator):
+    a, b = parse(a, b)
+    return operate[operator](a, b)
+
+def inorder(exp, ds):
+    if(exp.key == "identifier"):
+        return ds[exp.args['this']]
+
+    if(exp.key == "literal"):
+        return exp.args['this']
+
+    args = exp.args
+
+    left = args['this']
+    right = None
+    if 'expression' in args:
+        right = args['expression']
+
+    left_sol = inorder(left, ds)
+
+    right_sol = None
+    if right is not None:
+        right_sol = inorder(right, ds)
+
+    if right_sol is not None:
+        return evaluate(left_sol, right_sol, exp.key)
+    else:
+        return left_sol
+
+def apply_group_by(fields, ds: xr.Dataset, agg_funcs):
+    grouped_ds = ds
+    for field in fields:
+        field_parts = field.split("_")
+        groupby_field = field_parts[0]
+        if len(field_parts) > 1:
+            groupby_field = f"{groupby_field}.{field_parts[1]}"
+            groups = grouped_ds.groupby(groupby_field)
+            grouped_ds = apply_aggregation(groups, list(agg_funcs.values())[0])
```

Comment:
This logic is for grouping by time.month or time.day fields
But this would break things if someone tries to group by any field that has an underscore. forecast_time or pressure_levels.
It's better to validate for special fields that we are offering (like time.day, time.month, etc) and the user should still be able to groupby on fields that have an underscore.

Reply:
Will add as an enhancement later.